spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -126,7 +126,7 @@ abstract class SparkWrite implements Write, RequiresDistributionAndOrdering {
     this.writeSchema = writeSchema;
     this.dsSchema = dsSchema;
     this.extraSnapshotMetadata = writeConf.extraSnapshotMetadata();
-    this.partitionedFanoutEnabled = writeConf.fanoutWriterEnabled();
+    this.useFanoutWriters = writeConf.fanoutWriterEnabled() && !writeRequirements.hasOrdering();
```

Comment:
This is similar to what we did for merge-on-read writes.

```
boolean fanoutEnabled = context.fanoutWriterEnabled();
boolean inputOrdered = context.inputOrdered();
long targetFileSize = context.targetDataFileSize();

if (table.spec().isPartitioned() && fanoutEnabled && !inputOrdered) {
  return new FanoutDataWriter<>(writers, files, io, targetFileSize);
} else {
  return new ClusteredDataWriter<>(writers, files, io, targetFileSize);
}
```

Reply:
I will need to check our compaction, however.

Reply:
I'm trying to think through the implication of this change.

This assumes that if an ordering is present in `writeRequirements` then that ordering clusters data by partition. When clustered by partition, there's no need to use a fanout writer. I think it's always the case that an ordering will cluster by partition when generated by Iceberg, but this is not a guarantee of `SparkWriteRequirements` so it's a little risky.

Another implication is that `fanoutWriterEnabled` can _disable_ fanout writes but not _enable_ fanout writes as an override, which is a behavior change. If there is a case where incoming data isn't clustered for some reason -- whether that's a bug in Iceberg or a problem in Spark -- then there is no way to override the behavior and use fanout to fix the write. I think that's a problem because it would block writes to a table.

I think that this PR should instead make `SPARK_WRITE_PARTITIONED_FANOUT_ENABLED` optional. When present the explicit value should be used. When not present, then the logic should default to `!writeRequirements.hasOrdering`.